A média das médias
==================



## Objetivo



Crie ou obtenha um conjunto de dados numéricos que representa uma variável aleatória (certifique-se que seu conjunto de dados tem pelo menos 100 itens). Considere que este conjunto de dados é a sua população. Compute e reporte a média desta população.

Realize o processo de amostragem aleatória sem reposição neste conjunto de dados 100 vezes. Cada vez que realizar a amostragem, colete um certo número de dados (algo entre 25% e 75% dos dados, escolha o valor que preferir), compute a média das suas amostras e registre estas médias em uma lista.

Compute a média da lista de médias que obteve e compare este valor com a média da população. Discuta seu resultado.



## Resultado e discussão



#### Criando o conjunto de dados

In [3]:
from numpy.random import choice
import numpy as np

cnj_dados = np.random.randint(0,1000, (20,20))
#C = [1,2,3,4,5,6,7,8,9,10]

media = cnj_dados.mean()
print(f'A média da população é: {media}')

A média da população é: 526.37


#### Amostragem

In [93]:
tam1 = 100
tam2 = 300
pop = cnj_dados.flatten()

amostra1 = choice(pop, tam1, replace=False)
media1 = amostra1.mean()
print(f'A estimativa da média da população com uma amostra de 25% é: {media1}')

amostra2 = choice(pop, tam2, replace=False)
media2 = amostra2.mean()
print(f'A estimativa da média da população com uma amostra de 75% é: {media2}')

A estimativa da média da população com uma amostra de 25% é: 505.67
A estimativa da média da população com uma amostra de 75% é: 499.93333333333334


#### 100 amostras

In [37]:
dados = np.random.randint(0,1000, (25,4))
pop1 = dados.flatten()

#A fração (frac) deve ser um número inteiro, exemplo: 75% da população, escrever '75'
# qtd significa quantidade da amostras que quer produzir
def amostragem(lista, total, frac, qtd):
    tam = (total * frac) * (100 ** (-1))
    dic = {
        "media da população": [],
        "amostras": [],
        "media das amostras": [],
        "media das medias das amostras": [],
        "sublistas": [],
    }
    dic["media da população"].append(lista.mean())
    for i in range(qtd):
        am = choice(lista, int(tam), replace=False)
        dic["amostras"].append("M" + str(i))
        dic["sublistas"].append(am)
        dic["media das amostras"].append(round(am.mean() , 1))
        dic["media das medias das amostras"] = round(np.mean(dic["media das amostras"]),1)
    return dic

##### 75% da população

In [42]:
x = amostragem(pop1, 50, 75, 100)
print(f'A média da população é : {x["media da população"]}')
print()
#print(f'A estimativa de média da população com 75% de amostragem é : {x["media das amostras"]}')
#print()
print(f'A média da média das amostras é : {x["media das medias das amostras"]}')

A média da população é : [485.65]

A média da média das amostras é : 483.8


##### 25% da população

In [41]:
y = amostragem(pop1, 50, 25, 100)
print(f'A média da população é : {y["media da população"]}')
print()
#print(f'A estimativa de média da população com 75% de amostragem é : {y["media das amostras"]}')
#print()
print(f'A média da média das amostras é : {y["media das medias das amostras"]}')

A média da população é : [485.65]

A média da média das amostras é : 490.9


Nesse caso, por conta da função ter sido feita para criar um único dicionário com a informação de 100 amostras para 1 única porcentagem por vez, ela não é útil para fins de comparação. Note que, algumas vezes, 25% da média das amostras é mais preciso que 75%, o que não é algo observado tão facilmente, ainda mais quando as amostras são pegas de forma aleatória e em porcentagens diferentes.
Sabendo disso, criamos uma função que resulta em 2 dicionários, podendo escolher 2 porcentagens. Nesse caso, as 200 amostras com proporções diferentes podem ser comparadas, demonstrando a real diferença de precisão.

### Comparação das amostras da população

In [28]:
#Frac1 e 2 determinam porcetagens diferentes para serem comparadas
def amostragem_comp(lista, total, frac1, frac2, qtd):
    tam11 = (total * frac1) * (100 ** (-1))
    tam22 = (total * frac2) * (100 ** (-1))
    dic1 = {
        "media da população": [],
        "amostras": [],
        "media das amostras": [],
        "media das medias das amostras": [],
        "desvio padrão da media": [],
        "desvio²": [],
        "sublistas": [],
    }
    dic2 = {
        "media da população": [],
        "amostras": [],
        "media das amostras": [],
        "media das medias das amostras": [],
        "desvio padrão da media": [],
        "desvio²": [],
        "sublistas": [],
    }
    dic1["media da população"].append(lista.mean())
    dic2["media da população"].append(lista.mean())
    for i in range(qtd):
        am1 = choice(lista, int(tam11), replace=False)
        dic1["amostras"].append("M" + str(i))
        dic1["sublistas"].append(am1)
        dic1["media das amostras"].append(round(am1.mean() , 1))
        dic1["media das medias das amostras"] = round(np.mean(dic1["media das amostras"]),1)
        dic1["desvio padrão da media"].append(round(np.std(dic1["media das amostras"])))
    for i in range(qtd):
        am2 = choice(lista, int(tam22), replace=False)
        dic2["amostras"].append("M" + str(i))
        dic2["sublistas"].append(am2)
        dic2["media das amostras"].append(round(am2.mean() , 1))
        dic2["media das medias das amostras"] = round(np.mean(dic2["media das amostras"]),1)
        dic2["desvio padrão da media"].append(round(np.std(dic2["media das amostras"])))
    return dic1, dic2

In [29]:
dados = np.random.randint(0,1000, (25,4))
pop1 = dados.flatten()

doiscinco = 25
setecinco = 75

a, z = amostragem_comp(pop1, 50, doiscinco, setecinco, 100)
print(f'A média da população é : {z["media da população"]}')
print()
#print(f'A estimativa de média da população com {doiscinco}% de amostragem é : {a["media das amostras"]}')
#print(f'A estimativa de média da população com {setecinco}% de amostragem é : {z["media das amostras"]}')
#print()
print(f'A média da média das amostras de {doiscinco}% é : {a["media das medias das amostras"]}')
print(f'A média da média das amostras de {setecinco}% é : {z["media das medias das amostras"]}')
print()
print(f'O desvio padrão da média das amostras de {doiscinco}% é : {a["desvio padrão da media"]}')
print(f'O desvio padrão da média das amostras de {setecinco}% é : {z["desvio padrão da media"]}')

A média da população é : [491.33]

A média da média das amostras de 25% é : 499.5
A média da média das amostras de 75% é : 493.3

O desvio padrão da média das amostras de 25% é : [0, 57, 51, 63, 62, 75, 70, 71, 70, 71, 73, 71, 68, 67, 66, 68, 72, 70, 68, 69, 69, 68, 69, 80, 78, 77, 77, 78, 80, 80, 79, 78, 79, 78, 77, 77, 76, 77, 76, 82, 81, 80, 80, 83, 82, 83, 84, 85, 84, 84, 83, 85, 85, 85, 85, 84, 86, 86, 85, 85, 84, 84, 83, 83, 82, 82, 81, 81, 80, 81, 81, 80, 80, 80, 80, 80, 80, 80, 80, 79, 79, 79, 79, 79, 79, 79, 80, 80, 80, 80, 80, 80, 79, 79, 78, 80, 80, 80, 80, 80]
O desvio padrão da média das amostras de 75% é : [0, 2, 17, 33, 34, 31, 31, 36, 38, 37, 37, 35, 34, 33, 32, 31, 36, 35, 35, 38, 41, 40, 40, 40, 40, 41, 43, 43, 42, 44, 44, 44, 43, 44, 44, 44, 44, 44, 44, 44, 44, 43, 43, 43, 42, 42, 42, 43, 43, 42, 43, 43, 43, 43, 42, 43, 43, 44, 43, 44, 44, 43, 43, 43, 43, 43, 43, 42, 42, 43, 43, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 41, 42, 42, 42, 42, 42, 42, 42, 43, 43, 42, 43, 4

#### Discussão
É visivelmente observável que a estimativa da média da população se aproxima cada vez mais do valor real quando se considera uma porcentagem de amostragem maior. Por exemplo, em um dos casos rodados com amostras aleatórias, a média da população foi 485.65, para 2 estimativas de 75 e 25 respectivamente: 488 e 476.9. Em uma análise simples, é visível que a amostragem maior tem uma estimativa mais fiel ao ideal. 
Além disso, é possível analisar pelo cálculo feito do desvio padrão, com intuito de realmente notar quanto os conjuntos de amostras variaram em relação às suas média, e nele o desvio padrão indica que para cerca de 40 unidades de desvio padrão o conjunto de amostras de 75%, o conjunto de 25% varia o dobro, em média sendo 80 UN.

Uma observação curiosa é que a amostragem da população pode aparentar ter valores repetidos, se em algum momento isso for observado (é mais fácil ver com uma população be
m menor), ainda que o comando replace seja equivalente a False. Isso acontece porque o cnj_dados pode ter entre os valores randomizados, alguns valores que sejam iguais, porém em posições diferentes. Por conta disso, podem aparecer valores iguais, mas de posições diferentes no array que foi criado. Se fosse usado um conjunto C qualquer com valores não repetidos, isso não aconteceria.

## Referências



Escreva nesta seção as referências consultadas.

https://stackoverflow.com/questions/25440008/python-pandas-flatten-a-dataframe-to-a-list

http://www.bosontreinamentos.com.br/programacao-em-python/como-arredondar-numeros-em-python/

https://stackoverflow.com/questions/25440008/python-pandas-flatten-a-dataframe-to-a-list

https://acervolima.com/amostragem-aleatoria-em-numpy-funcao-sample/

## Playground

